packages/mcdoc/src/validator/index.ts

```diff
@@ -0,0 +1,896 @@
+	| SimplifiedStructType
+	| TupleType
+
+export interface SimplifiedEnum extends EnumType {
+	enumKind: EnumKind
+}
+export interface SimplifiedStructType extends StructType {
+	fields: SimplifiedStructTypePairField[]
+}
+export interface SimplifiedStructTypePairField extends StructTypePairField {
+	key: SimplifiedMcdocTypeNoUnion
+}
+
+const attributeHandlers: {
+	[key: string]:
+		| (<N>(
+			node: N,
+			attribute: Attribute,
+			inferred: McdocType,
+			expected: McdocType,
+			options: ValidatorOptions<N>,
+		) => ValidationError<N>[])
+		| undefined
+} = {
+	// TODO other attributes
+	id: (_n, attribute, inferred, _e, options) => {
+		if (inferred.kind === 'literal' && inferred.value.kind === 'string') {
+			// TODO check registry
+			if (!inferred.value.value.includes(':')) {
+				inferred.value.value = 'minecraft:' + inferred.value.value
+			}
+		}
+		return []
+	},
+}
+
+export function reference<T>(
+	node: T,
+	path: string,
+	options: ValidatorOptions<T>,
+) {
+	typeDefinition(node, { kind: 'reference', path: path }, options)
+}
+
+export function dispatcher<T>(
+	node: T,
+	registry: FullResourceLocation,
+	index: string | Index | ParallelIndices,
+	options: ValidatorOptions<T>,
+) {
+	const parallelIndices: ParallelIndices = typeof index === 'string'
+		? [{ kind: 'static', value: index }]
+		: Array.isArray(index)
+		? index
+		: [index]
+	typeDefinition(node, {
+		kind: 'dispatcher',
+		registry: registry,
+		parallelIndices: parallelIndices,
+	}, options)
+}
+
+export function typeDefinition<T>(
+	node: T,
+	typeDef: McdocType,
+	options: ValidatorOptions<T>,
+) {
+	const inferredType = options.inferType(node)
+	const errors = validate(node, typeDef, inferredType, options, [])
+
+	reportErrors(errors, options)
+}
+
+export function isAssignable(
+	typeDef: McdocType,
+	other: McdocType,
+	ctx: CheckerContext,
+	isEquivalent?: NodeEquivalenceChecker,
+): boolean {
+	const options: ValidatorOptions<McdocType | SimplifiedStructTypePairField> =
+		{
+			context: ctx,
+			inferType: d =>
+				d.kind === 'pair' ? { kind: 'tuple', items: [d.key, d.type] } : d,
+			isEquivalent: isEquivalent ? isEquivalent : () => false,
+			getChildren: d => {
+				if (d.kind === 'pair') {
+					return [d.key, d.type]
+				}
+				const simplified = simplify(d, d, d, options, [])
+				switch (simplified.kind) {
+					case 'list':
+						return [simplified.item]
+					case 'byte_array':
+						return [{ kind: 'byte' }]
+					case 'int_array':
+						return [{ kind: 'int' }]
+					case 'long_array':
+						return [{ kind: 'long' }]
+					case 'struct':
+						return simplified.fields
+					case 'tuple':
+						return simplified.items
+					default:
+						return []
+				}
+			},
+			getRange: () => Range.Full,
+			attachTypeInfo: () => {},
+		}
+
+	return validate(typeDef, other, typeDef, options, []).length === 0
+}
+
+function validate<T>(
+	node: T,
+	typeDef: McdocType,
+	inferredType: McdocType,
+	options: ValidatorOptions<T>,
+	parents: T[],
```

Comment:
`AstNode`s already have a `parent` field, so I think it's unnecessary and messier to pass the parent path along this recursive function. The error reporting should be able to get the correct parent from the node itself.

Reply:
T may not be an AstNode.

As an example, we could compare two McdocTypes to see if they are assignable.